### ISSUE TITLE
Request: Add Access-Control-Allow-Origin to include/http_header.h

### ISSUE THREAD

**Issue by cyremur** [pistacheio/pistache#92]
When combining a pistache backend server with interchangable javascript frontends that may be hosted on a different URL, one needs to set `Access-Control-Allow-Origin` origin headers to allow access to resources. 

Since this is a rather common and useful header, I would like to request adding it to the default headers implemented in the standard pistache build.

Alternatively, it would be nice to be able to have a simple way to add "raw" text headers to response. I tried `response.headers().addRaw(Http::Headers::Raw("Access-Control-Allow-Origin", "*"))` 
but that didn't seem to be actually written to the output and I am not completely sure what `addRaw()` is supposed to do.

**Comment by oktal** [comment]
Hello,

I initially added raw headers so that parsing would not fail when encountering a header not known to the system.

Raw headers could also be used for your particular use-case. However, raw headers are currently not serialized into the final http response sent back to the client.

I guess this is something we could also add so that you could add any header you want, even though it's not built-in into pistache. You would lose the type-safety aspect but that's a trade-of to make

**Comment by cyremur** [comment]
Hi,
So for this project I just added the class `AccessControlAllowOrigin`. 20 lines of code for a simple KVP header boilerplate is actually not that bad so raw headers are definitely not a priority and probably not even needed at all.

What's your stance on including more headers into the pistache repository? Should we just PR every new header we implement in order to expand standard pistache?

On another note, is there a way to set headers for multiple URLs sharing the same basepath or description? I need to set some headers, i.e. the access control headers, globally for many URLs and it would be very convenient if you could just set headers at description level.

**Comment by oktal** [comment]
> What's your stance on including more headers into the pistache repository? Should we just PR every
> new header we implement in order to expand standard pistache?

Yes, if the header you want to add is somewhat standard, feel free to submit a PR.

> On another note, is there a way to set headers for multiple URLs sharing the same basepath or description?
> I need to set some headers, i.e. the access control headers, globally for many URLs and it would be very convenient if you could just set headers at description level.

Currently, you need to add headers to every route. So you're saying that all your routes are sharing the same sets of headers and you'd like to add your set of headers at the description level so that you don't need to repeat them every-time you add a route ?

In your case, I think I'll considering creating a function that does that for you, i.e you tell it the route you want to add and it automatically adds the route to the description plus your headers.

API-wise, I'm still not sure if we should provide such a facility, feel free to discuss that

**Comment by cyremur** [comment]
> Currently, you need to add headers to every route. So you're saying that all your routes are sharing the same sets of headers and you'd like to add your set of headers at the description level so that you don't need to repeat them every-time you add a route ?

Yes, currently I want to add an AccessControlAllowOrigin header to every single route, which means every single function that handles a request starts with `response.headers().add<Http::Header::AccessControlAllowOrigin>("*");`.
(I haven't dug deep enough into the `route` code yet to find out other ways to add headers to responses, and this was the easiest possibility that worked for me.)
Personally, I would much rather have something like 
`void createDescription() {
  desc
    .headers().add<AccessControlAllowOrigin>("*");
}`
which writes the header to a standard set of headers that are automatically attached to every response from the description. 

Basically, I'm imagining the routes as tree with routes as leafs, the description as the root, and basepaths as inner nodes. I think it would be useful and neat to be able to attach headers at any point of the tree. A response would then collect every header on the path from leaf to route, with headers closer to the leaf taking precedence so you can override standard options if need be.

Sorry for blowing up a side question into such a feature discussion. I realize that the use-cases might not be pressing enough to warrant the amount of work going into a feature as described above, but I just really like the work you've done on pistache so far and got a little invested into the framework.

**Comment by oktal** [comment]
Hello,

> Basically, I'm imagining the routes as tree with routes as leafs, the description as the root, and basepaths as inner nodes. I think it would be useful and neat to be able to attach headers at any point of the tree.

Yes, that makes perfect sense. As of today, if we want to support such a feature, I think that the description code must be re-factored a little bit but this is definitely a good feature request

**Comment by oktal** [comment]
Merged in a3af796e7bc16886c50ada4e8dfadfca97080863

**Comment by xoac** [comment]
I think this can by closed /close
